### PR TITLE
Shrink entry sizes to reduce memory and IO usage

### PR DESCRIPTION
**Pull request by rostislav**
Efficient plotting of k=32 with 2 threads is now possible with buffer size of 3389 MiB.

### CODE REVIEW

**Comment by mariano54** [comment]
This is amazing! I'll try to review to see if all edge cases work. We should also run a few 32s to test (and a 33/34 in case the issue shows up on a higher k).

**Review by arvidn — COMMENTED** [review]
looks good! I think we need a pretty good testing story before we land this. Just running and validating a few k32s, is that really enough to feel confident this is correct?

afaict, if we overflow, we just produce corrupt data, and I would expect that checking a plot is probabilistic, so it would still be possible to not discover an error.

File: src/phase2.hpp [hunk 1]
```diff
@@ -52,14 +52,15 @@ Phase2Results RunPhase2(
     uint32_t const log_num_buckets,
     bool const show_progress)
 {
-    // An extra bit is used, since we may have more than 2^k entries in a table. (After pruning,
-    // each table will have 0.8*2^k or fewer entries).
+    // After pruning each table will have 0.865 * 2^k or fewer entries on
+    // average
```
[the remark this answers is not in the document]
Comment (rostislav) [reply src/phase2.hpp:56]: I observed empirically that the number of entries in tables 1-6 closely approximates the theoretical expected values. I also did the math to confirm that the probability of exceeding 2^k entries in those tables is negligible. Below is a calculation for table 6:

Assuming the table has `n = 2^k` entries after phase 1 and that `2^k` random pairs of entries match, the probability of an entry _not having_ a match is `((n-1)/n) ** (2*n) = (1 - 1/n) ** (2*n)`, which is approximated by `e^(-2) = 0.135` for a sufficiently large `n`. The probability of an entry _having_ a match (and thus remaining in table 6) is `p = 1 - 0.135 = 0.865`.

Now we can use the central limit theorem to prove that for a sufficiently large `n`, the number of matching entries `m` can be approximated by a normal distribution with `mean = n * p = 2^k * 0.865` and `stdev = sqrt(n * p * (1 - p)) = sqrt(2^k * 0.865 * (1 - 0.865)) = 2^(k/2) * 0.342`. The number of standard deviations required to reach `2^k` is given by `d = (2^k - mean) / stdev = 2^k * 0.135 / (2^(k/2) * 0.342) = 2^(k/2) * 0.396`. For the minimal `k = 18` we have `d = 2^9 * 0.395 = 202`, and it grows exponentially with larger `k`. It is clear that the probability of deviating from the mean by more than 200 sigma is so small that it can be considered impossible.

The number of entries in tables 1-5 after pruning is smaller than in table 6, and similar calculations also apply.

File: src/phase3.hpp [hunk 2]
```diff
@@ -131,6 +131,7 @@ Phase3Results RunPhase3(
     const bool show_progress)
 {
     uint8_t pos_size = k;
+    uint8_t line_point_size = 2 * k - 1;
```
[the remark this answers is not in the document]
Comment (rostislav) [reply src/phase3.hpp:134]: OK, will fix.

File: src/entry_sizes.hpp [hunk 3]
```diff
@@ -60,16 +60,23 @@ class EntrySizes {
...
             default:
                 // Represents line_point, f7
-                return Util::ByteAlign(3 * k) / 8;
+                return Util::ByteAlign(3 * k - 1) / 8;
         }
     }
 
+    // Get size of entries containing (sort_key, pos, offset). Such entries are
+    // written to table 7 in phase 1 and to tables 2-7 in phase 2.
+    static uint32_t GetKeyPosOffsetSize(uint8_t k)
```
[the remark this answers is not in the document]
Comment (rostislav) [reply src/entry_sizes.hpp:75]: I made it `static` to be consistent with other functions in the same class. The compiler is likely to make it inline anyway.

[the remark this answers is not in the document]
Comment (arvidn) [reply src/entry_sizes.hpp:75]: oh, I see. this is a member function. I thought this was `static` as in internal linkage.
This class is really just a namespace.

**Review by wjblanke — APPROVED** [review]
aok
